arch/cortex-m0/src/lib.rs

```diff
@@ -191,21 +307,71 @@ pub unsafe extern "C" fn svc_handler() {
 
 // Mock implementation for tests on Travis-CI.
 #[cfg(not(any(target_arch = "arm", target_os = "none")))]
-pub unsafe extern "C" fn switch_to_user(
-    _user_stack: *const u8,
-    _process_regs: &mut [usize; 8],
-) -> *mut u8 {
+unsafe extern "C" fn hard_fault_handler() {
     unimplemented!()
 }
-
 #[cfg(all(target_arch = "arm", target_os = "none"))]
-#[no_mangle]
-pub unsafe extern "C" fn switch_to_user(
-    mut user_stack: *const u8,
-    process_regs: &mut [usize; 8],
-) -> *mut u8 {
+#[naked]
+unsafe extern "C" fn hard_fault_handler() {
     use core::arch::asm;
+    // If `kernel_stack` is non-zero, then hard-fault occurred in
+    // kernel, otherwise the hard-fault occurred in user.
     asm!("
+    /*
+     * Will be incremented to 1 when we determine that it was a fault
+     * in the kernel
+     */
+    movs r1, #0
+    /*
+     * r2 is used for testing and r3 is used to store lr
+     */
+    mov r3, lr
+
+    movs r2, #4
+    tst r3, r2
+    beq 100f
+
+// _hardfault_psp:
+    mrs r0, psp
+    b 200f
+
+100: // _hardfault_msp
+    mrs r0, msp
+    adds r1, #1
+
+200: // _hardfault_exit
+
+    b {}    // Branch to the non-naked fault handler.
+    bx lr   // If continued function returns, we need to manually branch to
+            // link register.
+    ",
+    sym hard_fault_handler_continued,
+    options(noreturn));
+}
+
+pub enum CortexM0Subarch {}
```

Comment:
I assume this is an enum rather than a struct so that it cannot be instantiated -- should we make a note about that in a comment?

Reply:
Yes, that's the idea. I'll add a comment :+1:.